### DISCUSSION TITLE
How to get individual viscosity and thermal conductivity of species

### DISCUSSION THREAD

**Discussion by sramjatan1** [mutationpp/Mutationpp#154] [Q&A]
Hello,

I pulled a boundary layer from a CFD simulation and want to calculate individual viscosity and thermal conductivity of some species.   I know the mass fraction of the species and have the temperature profile from the CFD simulation. 

I know mutation can give the viscosity and thermal conductivity of each species in a mixture by doing a Chapman-Enskog analysis, correct? Can I somehow pass an array of mole fraction and temperature profile and get the resulting individual viscosity and thermal conductivity of some species e.g.  N2,N and O? 

I appreciate any insight into this!

**Comment by jbscoggi** [comment]
Hi @sramjatan1, thanks for the question and being the first to use the discussion feature!

A couple of things I want to clarify:

    The Chapman-Enskog solution to the Boltzmann equation provides mixture transport coefficients in terms of the so-called bracket integrals which can be translated into a set of linear systems by use of a Galerkin method which can be solved using appropriate linear system solution techniques, including iterative or direct methods. These are referred to as multicomponent transport algorithms and require knowledge of species compositions and temperatures.
    If you are only interested in computing individual species transport coefficients, then you don't need to solve transport systems, but instead can evaluate them directly, normally as a function of temperature only. The composition of the mixture is not required, but I think that computing the individual properties has limited utility unless you really have a pure gas of a single species.

With that said, I will assume that what you really want is the mixture properties for a given set of species mass fractions and temperature and answer accordingly. If in fact you do want pure species properties, just let me know.

The short answer is that there is currently no tool shipped with M++ that can do this from the command line, but it is easy enough to use M++ to write a small C++ code to do it. I would refer you to the tutorial on getting started which can be easily modified to do what you need. You can modify the tutorial to load your mixture file and set the temperatures and densities (note that mass fractions and temperature is not enough to fully specify the thermodynamic state). Once the state is set, you can easily compute the thermal conductivity and viscosity as shown in the tutorial.

Please let me if something is not clear from the tutorial or if you really do want the single species properties.

**Reply by sramjatan1** [reply]
Hi JB,

I was reading the tutorial section and was trying to compile the tutorial.cpp. I was getting an error when compiling (attached) and realized I need to add the eigen dependency, correct? My question was how do I add this dependency? 

Thanks,

![error_Eigen3](https://user-images.githubusercontent.com/57545172/121748474-096f4a80-cad7-11eb-9c1d-782c09fc0b3a.PNG)

**Reply by rdbisme** [reply]
Hello @sramjatan1! 

Did you give a look to [the docs about dependencies](https://github.com/mutationpp/Mutationpp/blob/master/docs/dependencies.md)? 

Is there something not applicable to your case?

**Reply by sramjatan1** [reply]
Hi @rubendibattista ! Thanks for responding. Sorry I am new to github so these questions will be bad lol So to install the eigen dependency, which command am I supposed to run? Is it something like: 

git subrepo status --all

I was reading the docs about dependencies but was not sure.

Thanks

**Reply by rdbisme** [reply]
No problem! :)
>Mutationpp depends on Eigen and on Catch2 for testing. Generally speaking the suggested approach is to use these dependencies **installing them using the system package manager** (or contact your system administrator to make them available to you, for instance as environment module). If they're available in the system in standard locations (or their module is loaded in the current shell), they should be retrieved automatically by CMake (if it does not, it's probably a bug, so open an issue on Github).

On which system are you? Maybe it's better if you open a separate issue and follow the template that you'll find in the issue to help me helping you! So please open an issue regarding the dependency! :)

**Reply by sramjatan1** [reply]
Okay thanks @rubendibattista!

**Reply by sramjatan1** [reply]
> Hi @sramjatan1, thanks for the question and being the first to use the discussion feature!
> 
> A couple of things I want to clarify:
> 
> ```
> The Chapman-Enskog solution to the Boltzmann equation provides mixture transport coefficients in terms of the so-called bracket integrals which can be translated into a set of linear systems by use of a Galerkin method which can be solved using appropriate linear system solution techniques, including iterative or direct methods. These are referred to as multicomponent transport algorithms and require knowledge of species compositions and temperatures.
> If you are only interested in computing individual species transport coefficients, then you don't need to solve transport systems, but instead can evaluate them directly, normally as a function of temperature only. The composition of the mixture is not required, but I think that computing the individual properties has limited utility unless you really have a pure gas of a single species.
> ```
> 
> With that said, I will assume that what you really want is the mixture properties for a given set of species mass fractions and temperature and answer accordingly. If in fact you do want pure species properties, just let me know.
> 
> The short answer is that there is currently no tool shipped with M++ that can do this from the command line, but it is easy enough to use M++ to write a small C++ code to do it. I would refer you to the tutorial on getting started which can be easily modified to do what you need. You can modify the tutorial to load your mixture file and set the temperatures and densities (note that mass fractions and temperature is not enough to fully specify the thermodynamic state). Once the state is set, you can easily compute the thermal conductivity and viscosity as shown in the tutorial.
> 
> Please let me if something is not clear from the tutorial or if you really do want the single species properties.

Hi @jbscoggi @rubendibattista  I believe I need individual species transport coefficients. I just wanted to explain in more detail what I need to make sure I am going in the right direction. To clarify, in order to sample from a Chapman-Enksog distribution, I need to compute the heat flux and shear stress  for each species in the boundary layer at each point. In other words, I need to get ```Tau= (-2/3)*mu(i,j)*dv/dy(i)``` and ```q= -k(i,j)*dT/dy(i)```. Here mu is the individual species viscosity, dv/dy is the y-velocity gradient, k is the individual species thermal conductivity, and dT/dy is the temperature gradient. 

So from the CFD solution, I have the partial densities (O,N2, and N), temperature, and velocities at each point in the boundary layer.  [y |rho_N |rho_N2| rho_O| T| u| v|]

So from the CFD solution, I can get the temperature and velocity gradient. In order to get the shear stress and heat flux
in my DSMC simulation for a given species like oxygen ```Tau= (-2/3)*mu(i,j)*dv/dy(i)``` I need the viscosity for oxygen.

So I believe I need individual species transport coefficients, correct? 

Will I still need to pass in the partial densities and temperature array? I can follow the tutorial.cpp to get this done right?

Thanks

**Reply by jbscoggi** [reply]
Hi @sramjatan1, sorry but I'm still a little unclear about what you are doing.  Are the indices i and j in your equations related to the CFD mesh or are they species indices?  I guess I don't understand where the index in dv/dy(i) is coming from if it's not from the mesh.  In that case, it seems you want the mixture viscosity.  However, if you are trying to transfer the solution from CFD to a DSMC code, then I am really confused because I think the transport properties are only available (or meaningful) as a post processing quantity.  The DSMC collision operator requires collision cross-sections, not transport properties.  In fact, transport properties are really only defined in the continuum limit.  Could you provide a few more details about what you are trying to do, or perhaps provide a reference that could help explain?

**Reply by sramjatan1** [reply]
Hi @jbscoggi 

I sent you an email to try to clarify and will also try to clarify here.

Regarding tutorial.cpp in the examples folder, currently mixture properties are outputted for air_5
which consists of the species ``` N O NO N2 O2```. Now the example computes the ```Mixture thermal conductivities [W/m-K]``` and the
```Mixture dynamic viscosity [Pa-s]``` at 300 K and 1 ATM. 

Now lets say I want to know the individual species ``` ( N O NO N2 O2) ``` thermal conductivities and dynamic viscosities at 300 K and 1 ATM. 

How can I get extract this information?? Based on the equation below from your paper, here mu is the mixture quantity. How can I back out the individual viscosity value for lets say N2? 

![transport_mixture_quant](https://user-images.githubusercontent.com/57545172/123189476-ded4a880-d46b-11eb-8a8b-f53f2edc17ff.PNG)

**Reply by jbscoggi** [reply]
Hey @sramjatan1, sorry for not getting back to you by email, I have been very busy.  At any rate, I think it's better to keep the conversation here anyways in case others have the same problem to solve.  I thought about what you want and realized that you should be able to get this by calling the mppequil program with a pure gas for each species.  For example:

```bash
mppequil --species-list N -m 0,32,33
```

That will generate a table of viscosity and thermal conductivity for pure N versus temperature.  I guess it would not be too hard then to build this table for each of your species and do a simple linear interpolation to provide the properties at the temperature you need.  At least this way, you don't need to write/modify any code (except to interpolate the table).  It's also nice because if you change the library or method in the future, you only need to update the table with new values.  What do you think about this solution?

**Reply by sramjatan1** [reply]
@jbscoggi thanks for the reply! I see what your saying. A quick clarification question. When you use ```mppequil --species-list N -m 0,32,33``` which equation is being solved to get the viscosity as a function of temperature? Is it possible to please refer me to an equation from one of your mutation papers? Also, I am trying to understand why when we want individual species transport properties we do not need to solve transport systems but can evaluate them directly? I am just thinking that everything is coupled somehow and that we need to solve some kind of transport system to get the individual species transport properties. Thanks!
